### PR TITLE
Requiring "Programmed" condition in conformance tests

### LINKED ISSUE #1722: Add missing GatewayConditionType conformance test
As per https://github.com/kubernetes-sigs/gateway-api/issues/1454, 

> - [ ] Update the description for the "Ready" condition to note the stricter behavior and that it is now extended conformance on:
>  - [ ] ListenerConditionType
>  - [ ] GatewayConditionType

`ListenerConditionType` and `GatewayConditionType` updated the description but there are no conformance test yet.

Here is the requet for the conformance test of `GatewayConditionType`.

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind test
/area conformance

**What this PR does / why we need it**:
This adds a requirement in conformance tests that Gateways must be "Programmed" (previously this was just "Accepted"). Note that this only covers the top level "Programmed" condition, not each Listener.

**Which issue(s) this PR fixes**:
Fixes #1722

**Does this PR introduce a user-facing change?**:
```release-note
Conformance: Gateways must publish the "Programmed" condition.
```

### CODE REVIEW

**Comment by robscott** [comment]
I believe that every time we make this check in conformance tests, we're also expecting the Gateway(s) to be programmed. If I've missed anything, let me know. I'm going to put a hold on this because I'd like to get sign off from a few people before merging. I know this will _completely_ break conformance tests for some implementations because it's a change at such a high level.

/hold
/cc @arkodg @howardjohn @mlavacca @skriss @sunjayBhatia

**Comment by arkodg** [comment]
@robscott from a quick glance it looks like this case is being addressed in Envoy Gateway [here](https://github.com/envoyproxy/gateway/blob/9785b4192ae47f5ed60d591deb654a2ca4155294/internal/status/conditions.go#L75) 
you could try testing it out using steps outlined [here](https://github.com/kubernetes-sigs/gateway-api/issues/1734#issuecomment-1432310808)

**Review by shaneutt — APPROVED** [review]
This seems like the right change to me, but definitely would like to invite some more people to check this out before we merge it.

File: conformance/utils/kubernetes/helpers.go
Comment (skriss) [inline conformance/utils/kubernetes/helpers.go:203]: This might be slightly more clear:
```suggestion
					t.Logf("%s/%s Gateway not accepted yet", ns, gw.Name)
```

File: conformance/utils/kubernetes/helpers.go
Comment (skriss) [inline conformance/utils/kubernetes/helpers.go:209]: Similarly:
```suggestion
					t.Logf("%s/%s Gateway not programmed yet", ns, gw.Name)
```

**Review by mlavacca — APPROVED** [review]
This change looks good to me 👍

**Review by skriss — COMMENTED** [review]
Thanks @robscott, this mostly looks good from Contour's perspective; unfortunately it's causing us to start failing the gateway-observed-generation-bump test because we don't yet support the addition of the second HTTP Listener port, but that's an issue for us rather than the conformance test.

File: conformance/utils/kubernetes/helpers.go
Comment (robscott) [reply conformance/utils/kubernetes/helpers.go]: Updated, thanks!

File: conformance/utils/kubernetes/helpers.go
Comment (robscott) [reply conformance/utils/kubernetes/helpers.go]: Updated, thanks!

**Comment by robscott** [comment]
Thanks everyone! Sounds like this is safe enough to remove the hold. Next LGTM will merge this.

/hold cancel

**Review by shaneutt — APPROVED** [review]
/lgtm

**Pull request merged**
